### ISSUE TITLE
Change how paragraphs are handled in Rich Text editor

### ISSUE THREAD

**Issue by laurent22** [laurent22/joplin#8600]
The current rich text editor supports both paragraphs (P tag) and lines separated by newlines (BR tags). The distinction between both can be confusing and [makes it hard to edit the text](https://blog.codinghorror.com/invisible-formatting-tags-are-evil/).

It is also not how other WYSIWYG editors, in particular Word, work. Paragraphs are simply blocks of texts separated by two newlines.

We should aim for something similar. It means that when rendering the Markdown, to the Rich Text editor in particular, we should use two BRs to separate paragraphs instead of a P tag. When pasting text we should do something similar.

Pressing Enter, should create a BR tag (not a P, as now).

When implementing this, we should however make sure it's not going to break:

- Existing user-created notes
- And Web Clipped notes (as HTML)

**Comment by ssotoen** [comment]
> It is also not how other WYSIWYG editors, in particular Word, work.

That’s exactly how Word works though. In Word you start a new paragraph with <kbd>Enter</kbd> and a new line with <kbd>Shift</kbd> + <kbd>Enter</kbd>, same as in Joplin.

**Comment by laurent22** [comment]
I've just tried from a blank document the other day and that's not how it was working

**Comment by tomasz1986** [comment]
I've just done a test in Word 2021 and this is how it looks:

![image](https://github.com/laurent22/joplin/assets/5626656/1b7de7e5-ccf6-4195-8796-d0f2b61a5324)

The "new line" was created using `Shift+Enter`, and the "new paragraph" using just `Enter`.

**Comment by laurent22** [comment]
I've just tried again on Word 2019 (macOS) and it's as I mentioned in the initial post. There's no "Shift+Enter", just Enter creates a new line, and Enter twice would be the equivalent of a new paragraph.

So now I'm confused, not sure what's best here.

<img width="317" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/5aecc590-eab9-4643-9c98-f8d7eab65ab7">

**Comment by tomasz1986** [comment]
I think there may be differences in how this works in Word on macOS and Word on Windows. The two aren't exactly the same, and especially the older versions (as from what I know, the newest ones have become closer in terms of functionality). My testing was done on Windows 10.

**Comment by Daeraxa** [comment]
This is dependent on what style you use. By default I think the `Normal` style is chosen which is the way that @tomasz1986 is describing but whenever I make a document I always swap to the `No Spacing` style to revert back to the more "traditional" mode.

Interestingly Wordpad also defaults to this same style with the different return and shift + return behaviour.

Compare that to LibreOffice Writer or Google Docs which both use the same for both.

**Comment by ssotoen** [comment]
Word, Google Docs, and Libreoffice Writer all work the same way. Enter for a paragraph break, Shift + Enter for a line break.
Google Docs and Writer’s default styles don’t include extra spacing so paragraph breaks look like line breaks, but the formatting marks show what’s actually going on.
¶ marks a paragraph break.
↵ marks a line break.

@laurent22 What does your document look like with formatting marks [enabled](https://support.microsoft.com/en-us/office/show-or-hide-tab-marks-in-word-84a53213-5d02-404a-b022-09cae1a3958b#ID0EBBD=macOS)?

![Screenshot-word2](https://github.com/laurent22/joplin/assets/68116836/473cd749-d64e-4f8a-9f0e-02bb862973f5)

![Screenshot-google](https://github.com/laurent22/joplin/assets/68116836/d745e167-0214-459e-8317-7a43144b5306)

![Screenshot-libreoffice](https://github.com/laurent22/joplin/assets/68116836/7386902d-9815-4dd4-9607-ef45540b7f06)

**Comment by laurent22** [comment]
With formatting enabled I have this:

<img width="255" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/339abd16-5938-4895-a72a-716b55bed0c0">

It doesn't matter if I press Shift+Enter or not, everything is just a basic newline.

With Google Docs I get the same behaviour:

<img width="148" alt="image" src="https://github.com/laurent22/joplin/assets/1285584/73f56f77-78e5-49d1-9032-c69310c723de">

It might indeed depends on the settings, but I rarely use Word (or Google Docs) so it's default settings everywhere.

**Comment by ssotoen** [comment]
Interesting. These are all paragraph breaks. Seems like Shift + Enter is broken on MacOS.

**Comment by laurent22** [comment]
Let's leave it as it is, since the current behaviour seems to be ok

**Comment by laurent22** [comment]
This issue comes up often enough that I'm wondering if we should have a poll about it.

**Comment by ghost** [comment]
Please add an option to change this. I don't want to press Shift + Enter every time to make a new line, I just want it to be Enter. I didn't see it in Keyboard Shortcuts.

I'm probably dumb but I don't understand what the point is with Viewer (I can't edit anything), and Markdown editor isn't pleasing either because I don't want to see the technical stuff when bolding or whatever, like `- **Note Editor**` instead of it just being bold and with a bullet point. Joplin with Rich Text Editor seems like just what I want, if only Enter would make a "normal" new line.

Sorry if it's just me that's being dumb and not understanding.

I want it like this:
![cet](https://github.com/laurent22/joplin/assets/124498221/16ff46f8-36de-4430-9821-037d77b077dd)

Not like this:
![cyt](https://github.com/laurent22/joplin/assets/124498221/752c2b4d-e60c-45f2-9977-6b81b59e430d)

**Comment by laurent22** [comment]
For that particular example, I guess you could use a list? That would be more semantically correct or would format in a better way.

But yes I agree that ideally pressing Enter should just be one new line, not a new paragraph. However I checked and the editor we use, TinyMCE, doesn't really support it. Well it does, but when it's enabled many things are broken because it needs the concept of paragraphs to apply formatting correctly.

So I can't think of a solution at this point. It seems like a Typora-like editor would solve this by treating the text as it is, instead of having hidden paragraphs or newline elements.

**Comment by personalizedrefrigerator** [comment]
One option would be to add a "add space between paragraphs" or similar appearance setting.

This could be applied with CSS similar to the following:
```css
#rendered-md.-reduce-paragraph-spacing p {
        margin-top: 0;
        margin-bottom: 0;
}
```

**Comment by artemgur** [comment]
I think that the best solution to paragraph vs new line on Enter debate is to make it a setting, so every user can choose an option that is more convenient to them.

For me, personally, paragraph on Enter is vastly preferred behavior (in fact, I never had the need to insert a "line but not paragraph" in the first place). And it's a de-facto standard in editors like Word. So the hypothetical change of that behavior would be workflow-breaking for me (and likely many others), and I would likely migrate to another notes app (because it would be extremely unconvenient to press Enter twice or Shift+Enter every time I want to start a new paragraph).

In my opinion, "just new line" is a rarer use case than "paragraph", so Shift+Enter is better for "just new line" and Enter for "paragraph". But I understand that others can have other opinion, so it would be best to let the user choose via settings.

This comment is only about the usability side of the issue, not about `<p>` vs `<br>`.

**Comment by softlypink** [comment]
Excuse my poor example. My point is not about a list, it's the different spacing compared to a 'regular' line, which just feels odd and not being able to click the 'lines' throws me off. I hope that makes sense. If I want a paragraph I will press Enter twice myself.

Anyway, I understand now and I thank you for explaining why it's difficult to change.

**Comment by captain118** [comment]
I too would like to have this as an option.  If you are wanting consistancy.  OneNote doesnt do it.  EverNote doesnt do it.  Obsidian doesnt do it.  Notepad++ doesnt do it.  It seems like word processors normally do this but Note taking apps dont normally do it.  I can appreciate people needing the option.  For me its more I would rather be able to see more of my notes on one screen than I would have the text be spaced out.

**Comment by captain118** [comment]
> Word, Google Docs, and Libreoffice Writer all work the same way. Enter for a paragraph break, Shift + Enter for a line break. Google Docs and Writer’s default styles don’t include extra spacing so paragraph breaks look like line breaks, but the formatting marks show what’s actually going on. ¶ marks a paragraph break. ↵ marks a line break.
> 
> @laurent22 What does your document look like with formatting marks [enabled](https://support.microsoft.com/en-us/office/show-or-hide-tab-marks-in-word-84a53213-5d02-404a-b022-09cae1a3958b#ID0EBBD=macOS)?
> 
> ![Screenshot-word2](https://private-user-images.githubusercontent.com/68116836/259492453-473cd749-d64e-4f8a-9f0e-02bb862973f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwODg4MTcsIm5iZiI6MTcxNjA4ODUxNywicGF0aCI6Ii82ODExNjgzNi8yNTk0OTI0NTMtNDczY2Q3NDktZDY0ZS00ZjhhLTlmMGUtMDJiYjg2Mjk3M2Y1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE5VDAzMTUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYmY2MGQzNjMwNzZkMjVjYzg5NzQ2YmRiNDRiNmNiYjM4OWE3NWZkNjE3ZjMyYWQxMGJjMTI0NmVmNGEzMjgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oDoJs144G3yAXlvKg32Dollxa-N7ZvhRHvH5v6KLqrU)
> 
> ![Screenshot-google](https://private-user-images.githubusercontent.com/68116836/259492518-d745e167-0214-459e-8317-7a43144b5306.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwODg4MTcsIm5iZiI6MTcxNjA4ODUxNywicGF0aCI6Ii82ODExNjgzNi8yNTk0OTI1MTgtZDc0NWUxNjctMDIxNC00NTllLTgzMTctN2E0MzE0NGI1MzA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE5VDAzMTUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ZWQwMWNlNWI1NGM3NjZmYmQ2NzYzY2YxMzlkNmQyMTQxNDNhMTRiODI2YmYwMjY4ZjI5MzQzYmM0ZDA3MWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.O-bg5-csARjlC7uE31LQHlA2JPbedDqsDnSJwPoUecI)
> 
> ![Screenshot-libreoffice](https://private-user-images.githubusercontent.com/68116836/259492488-7386902d-9815-4dd4-9607-ef45540b7f06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYwODg4MTcsIm5iZiI6MTcxNjA4ODUxNywicGF0aCI6Ii82ODExNjgzNi8yNTk0OTI0ODgtNzM4NjkwMmQtOTgxNS00ZGQ0LTk2MDctZWY0NTU0MGI3ZjA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE5VDAzMTUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3OTY0YWJjZTQ0YjU4ZDZhYjQ2MTA4ZTRkNTkxZDA5MWQ4MjRiNDg0ZjIwYzM0MzY4NWFhZDhhZGQwZGM4ODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WLtHAGVqyfhGlK8p0rkr9ZflSW-Cp5GBF6_8Fcv7lGU)

These are word processors not note taking apps.  They are a different catagory of application.

**Comment by steveb166** [comment]
I think I have to agree with adding it as a setting at least. My opinion is that its a note taking app not a word processor, I'm not sure I even see the point of a new "paragraph" when two new lines basically does the same thing when required, whilst avoiding the annoyance of pressing shift all the time.

**Comment by Australian-Infomaniac** [comment]
Yes **please please please please please provide an option to change it**.
It causes me problems every single day.
It's so disruptive, I've literally created a to do list item about it.
I've also searched for solutions about five times.
All the solutions create problems, like making all your _previously-created_ docs lose their spacing.
________________________________________
Every other program I've ever used just takes you to a new line when you hit enter, rather than adding a line/space/paragraph (whatever you may call it).
Various abstract arguments otherwise don't contradict this.
____________________________________________
Furthermore, since this is a note-taking program, I use new-paragraphs for specific purposes.
To create a whole new section.
_____________________________________________
Learning to hit shift-enter will just create a habit that confuses my use of other apps.
_____________________________________________
How about THIS as an option:
Some of us can choose a preference that creates a simple new line when you hit enter, 
and if you hit SHIFT enter, you can create a line-space.
Basically the reverse of now.
_____________________________________________
My number 1 wish with Joplin! :)

**Comment by viuxa** [comment]
Similar problems have been discussed by many people every year, I don't quite understand why developers are not willing to make changes. 

Joplin also has a lot of popular users, no matter non-technical or technical people, just a simple line break must Shift+Enter is still very annoying, and this problem used to unanimously bother me, then I used a simpler solution to solve the problem, shortcut mapping, on windows I use autohotkey to map shortcuts  Enter -> Shift+Enter, Shift+Enter -> Enter. 

I think if it's difficult to implement at least add the shortcuts to the settings, I subjectively think it's less disruptive and gives the user more choices.

**Comment by laurent22** [comment]
I'm copying my answer from the forum here and locking the thread so that this answer remains visible. We understand this is important to a lot of users (and, actually, to me too), but unfortunately there is no easy fix:

* * *

The problem is that if we change this behaviour to enter a newline instead of a paragraph, the editor will see the document as one giant continuous line with no paragraphs (even two lines separated by two newlines will **not** be considered as two paragraphs). So a lot of functionalities that rely on paragraphs will be broken, in particular when creating lists.

The conversion to Markdown also won't be good, which means when you view the note on mobile you'll have a very long line with litteral `<br>` tags to represent new lines.

So it's not really a matter of adding an option or not. Even if we add the option, it's still broken and unfortunately that can't currently be fixed with the editor we're using.

My preferred solution at the moment would be to move to a different Rich Text Editor, one like Typora, which is a bit closer to Markdown while keeping most of the advantages of a proper RTE.
